File: scalaz/src/main/scala/step0_repl.scala
import scala.annotation.tailrec
import scala.io.StdIn
import scala.util.control.NonFatal

object step0_repl {

  def read(str: String): String = str

  def eval(ast: String, env: String): String = ast

  def print(exp: String): String = exp

  def rep(str: String): String = print(eval(read(str), ""))

  def main(args: Array[String]): Unit = {
    @tailrec
    def go(): Unit = Option(StdIn.readLine("user> ")) match {
      case Some(line) =>
        try println(rep(line)) catch {
          case NonFatal(e) => e.printStackTrace()
        }
        go()
      case None => ()
    }

    go()
  }
}

File: java8/src/main/java/mal/step0_repl.java
package mal;

import java.io.IOException;

public class step0_repl implements Step<String, String> {

    public static void main(String[] args) throws IOException {
        step0_repl instance = new step0_repl();
        String line;
        while ((line = IO.prompt("user> ")) != null) {
            System.out.println(instance.rep(line, ""));
        }
    }

    @Override
    public String read(String str) {
        return str;
    }

    @Override
    public String eval(String ast, String env) {
        return ast;
    }

    @Override
    public String print(String ast) {
        return ast;
    }
}
